olm/operator-controller/v1alpha1/clusterextension_types.go

```diff
@@ -0,0 +1,518 @@
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/operator-framework/operator-controller/internal/conditionsets"
+)
+
+var ClusterExtensionKind = "ClusterExtension"
+
+type (
+	UpgradeConstraintPolicy string
+	CRDUpgradeSafetyPolicy  string
+)
+
+const (
+	// The extension will only upgrade if the new version satisfies
+	// the upgrade constraints set by the package author.
+	UpgradeConstraintPolicyCatalogProvided UpgradeConstraintPolicy = "CatalogProvided"
+
+	// Unsafe option which allows an extension to be
+	// upgraded or downgraded to any available version of the package and
+	// ignore the upgrade path designed by package authors.
+	// This assumes that users independently verify the outcome of the changes.
+	// Use with caution as this can lead to unknown and potentially
+	// disastrous results such as data loss.
+	UpgradeConstraintPolicySelfCertified UpgradeConstraintPolicy = "SelfCertified"
+)
+
+// ClusterExtensionSpec defines the desired state of ClusterExtension
+type ClusterExtensionSpec struct {
```

Comment:
The loose idea is that there are stages:
1. source the bundle from $somewhere (for now, from a catalog. in the future, perhaps from a direct reference or from a helm repo)
2. (not implemented yet) template/render the bundle using cluster-admin defined configuration
3. install it (i.e. change the cluster to reflect the result of (2))

So the idea was to have logical groupings of fields in the API so that fields from different concepts would not be interspersed.

Reply:
Just a note that a couple of the install fields have been promoted to the spec.

Reply:
Right, I wonder if the details of that flow can somehow be represented in the API field documentation so that it's easier to understand, but agreed, the logical grouping makes sense